ResolveComposesPlugin allows to handle FROM-scratch image

tkdchen opened the pull request:
* CLOUDBLD-127

Signed-off-by: Chenxiong Qi <cqi@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [x] Changes to metadata also update the documentation for the metadata
- [ ] Pull request has a link to an osbs-docs PR for user documentation updates
- [ ] New feature can be disabled from a configuration file

ben-alkov commented on atomic_reactor/plugins/pre_resolve_composes.py:
```diff
@@ -151,7 +151,13 @@ def adjust_for_autorebuild(self):
+        build_info = self.plugin_result.get(BASE_IMAGE_KOJI_BUILD)
+        if build_info is None:
+            self.log.debug('No base image koji build info is found. '
+                           'Skip to adjust inheritance for composes.')
```
@tkdchen; Based on this comment, I'm currently confused as to what's being skipped and why. Do you perhaps mean "Skipping forward to adjust inheritance for composes."?

tkdchen replied:
@ben-alkov Yeah, you are right. Sorry, "skip to" was used incorrectly.

athos-ribeiro commented:
/packit copr-build

tkdchen commented:
@athos-ribeiro 

The comment inside `InsideBuilder.__init__` mentions that if Dockerfile needs to be generated from `container.yml`, a plugin will create the dockerfile and set base image that sets `workflow.builder.base_from_scratch`. Which plugin does this job? Is this plugin optional to run for some cases before resolve_composes plugin?

athos-ribeiro commented:
> @athos-ribeiro
> 
> The comment inside `InsideBuilder.__init__` mentions that if Dockerfile needs to be generated from `container.yml`, a plugin will create the dockerfile and set base image that sets `workflow.builder.base_from_scratch`. Which plugin does this job? Is this plugin optional to run for some cases before resolve_composes plugin?

Currently, this is the case for Flatpak builds. There is a pre_create_flatpak_dokerfile plugin that does this. It rund before the resolve_composes plugin in the orchestrator build.

tkdchen commented:
@athos-ribeiro PR is updated and using `workflow.builder.base_from_scratch` now. PTAL.

athos-ribeiro commented:
LGTM!

The tests are checking for calls to log.debug, but there are other calls to `debug` in those methods. Could we improve those tests to either check for the logged strings or to check that the plugin will skip in the desired cases?

tkdchen commented:
@athos-ribeiro 

Thanks for your review. I updated this PR by checking the logged strings.

BTW, I also moved the check of `workflow.builder.base_from_scratch` to the front of the check of `self.plugin_result`. This is not for making the tests pass. My understand is if an image is based on scratch, the code should be actually unnecessary to consider the result from previous plugin, the koji_parent, because the `BASE_IMAGE_KOJI_BUILD` is the only thing `adjust_for_inherit` and `adjust_signing_intent_from_parent` care about.

athos-ribeiro commented:
/packit copr-build

athos-ribeiro commented:
Waiving osbs-docs PR and feature toggling through configuration requirements. This is a fix for a bug in the OSBS build workflow.

rcerven commented on tests/plugins/test_resolve_composes.py:
```diff
@@ -1365,3 +1366,21 @@ def get_override_yum_repourls(self, workflow, arch=ODCS_COMPOSE_DEFAULT_ARCH):
+
+    def test_skip_adjust_inheritance_for_composes_if_image_is_based_on_scratch(self, workflow):
+        plugin = ResolveComposesPlugin(workflow.builder.tasker, workflow)
+        flexmock(plugin.log).should_call('debug').once().with_args(
```
can we use just caplog? and check text in caplog.text

rcerven commented on tests/plugins/test_resolve_composes.py:
```diff
@@ -1365,3 +1366,21 @@ def get_override_yum_repourls(self, workflow, arch=ODCS_COMPOSE_DEFAULT_ARCH):
+            'This is a base image based on scratch. '
+            'Skipping forward to adjust inheritance for composes.'
+        )
+        flexmock(plugin.workflow.builder, base_from_scratch=True)
```
we can just set it plugin.workflow.builder.base_from_scratch=True
we don't need flexmock for that

rcerven commented on tests/plugins/test_resolve_composes.py:
```diff
@@ -1365,3 +1366,21 @@ def get_override_yum_repourls(self, workflow, arch=ODCS_COMPOSE_DEFAULT_ARCH):
+
+    def test_skip_adjust_signing_intent_from_parent_if_image_is_based_on_scratch(self, workflow):
+        plugin = ResolveComposesPlugin(workflow.builder.tasker, workflow)
+        flexmock(plugin.log).should_call('debug').once().with_args(
```
use caplog

rcerven commented on tests/plugins/test_resolve_composes.py:
```diff
@@ -1365,3 +1366,21 @@ def get_override_yum_repourls(self, workflow, arch=ODCS_COMPOSE_DEFAULT_ARCH):
+            'This is a base image based on scratch. '
+            'Signing intent will not be adjusted for it.'
+        )
+        flexmock(plugin.workflow.builder, base_from_scratch=True)
```
no need for flexmock

rcerven commented on atomic_reactor/plugins/pre_resolve_composes.py:
```diff
@@ -149,8 +149,14 @@ def adjust_for_autorebuild(self):
     def adjust_for_inherit(self):
+        if self.workflow.builder.base_from_scratch:
+            self.log.debug('This is a base image based on scratch. '
+                           'Skipping forward to adjust inheritance for composes.')
```
'Skipping forward to adjust inheritance for composes.' =>
'Skipping adjusting composes for inheritance'

tkdchen commented:
@rcerven Thanks for your good points. All your comments are addressed. PTAL.

athos-ribeiro commented:
/packit copr-build

athos-ribeiro commented:
/packit copr-build

rcerven commented:
/packit copr-build

chmeliik commented:
Release note: resolve_composes plugin will correctly handle FROM-scratch images
